### DISCUSSION TITLE
exllamav2 and IBM Granite models

### DISCUSSION THREAD

**Discussion by RaRasputinRGLM** [turboderp/exllamav2#460] [General]
I wanted to work with these because the sizes are appealing for local LLMs, they all use the starcoder tokenizer, however the model.embed_tokens is not found in any of the safe tensors, I was thinking, that I could hack a solution out with extracting the safetensor piece from the starcoder, but I think there might be a faster way to go that I am missing due to inexperience. 

I don't think this is really a exllamav2 issue but worth discussing. Here is a similar issue with IBM granite that can give some insight about how they structured their models, over at llama.cpp https://github.com/ggerganov/llama.cpp/issues/7116

In short what is the right route with exllamav2 and missing embed_tokens if you know the tokenizer it uses?

**Comment by turboderp** [comment]
I'm unsure what the issue is? I added support for Granite two weeks ago.

It indeed doesn't have an `embed_tokens` tensor, but that's just because GPTBigCode calls it `wte` instead. It has a number of other features that set it apart from Llama, like learned embeddings, non-gated MLP and bias everywhere, but that should all be accounted for.

**Reply by RaRasputinRGLM** [reply]
I am trying to quantize the instruct model https://huggingface.co/ibm-granite/granite-20b-code-instruct while building the measurement.json is throwing the error model.embed_tokens not found. I am using this command 

python convert.py
     -i /mnt/LLMs/granite-20b-code-instruct/ \
     -o /mnt/LLMs/tmp/ \
     -nr \
     -om /mnt/LLMs/exl_models/granite20b-exl2/measurement.json 

I have not tried on the base model waiting for it to download

It throws this error:
Traceback (most recent call last):
  File "/home/RGLM/projects/exllamav2/convert.py", line 70, in <module>
    config.prepare()
  File "/home/RGLM/projects/exllamav2/exllamav2/config.py", line 316, in prepare
    raise ValueError(f" ## Could not find {prefix}.* in model")
ValueError:  ## Could not find model.embed_tokens.* in model

it doesn't seem to be grabbing the gpt2_keymap that are set in the arch ExLlamaV2ArchParams

edit: added error and idea

**Reply by RaRasputinRGLM** [reply]
I tracked it down to the init for the STfile in fastensors.py, the header key is transformer.wte.weight
instead of just wte.weight, so it never sees it since it isn't the beginning of the header key

**Reply by turboderp** [reply]
That appears top be a new bug that's crept in. I did convert that model successfully here: https://huggingface.co/turboderp/granite-20b-code-instruct-exl2

I'll fix it in a moment.

**Reply by turboderp** [reply]
There, fixed it. It crept in while I was refactoring and adding GPT2 support. If you pull the latest master it should allow you to quantize. Inference on quantized models should work even without the fix (with a prebuilt wheel of 0.0.21 for instance) since the tensors are remapped as part of the conversion.

**Reply by RaRasputinRGLM** [reply]
Thanks for the fix, learned more about these models doing this, it was a much easier fix than the route I was going down.

**Reply by turboderp** [reply]
Thanks for pointing it out. Probably wouldn't have caught it for a while.
